Promote NU3043 warning to error in dotnet nuget sign command

kartheekp-ms opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->

# Bug

<!-- If this is an engineering change or test change only, you do not need an issue. -->
<!-- Find or create an issue in NuGet/Home and paste the full url. -->
<!-- At the maintainers discretion, multiple changes may apply to a single issue, but only when the PRs are all created within a short period of time. -->
Fixes: https://github.com/NuGet/Home/issues/13814

## Description

Updated dotnet nuget sign command to accept fingerprints from the SHA-2 family (SHA256, SHA384, or SHA512) instead of SHA-1. If a SHA-1 fingerprint or invalid value is passed, the commands raise an [NU3043](https://learn.microsoft.com/nuget/reference/errors-and-warnings/nu3043) error indicating that SHA-1 is insecure. 

## PR Checklist

- [x] Meaningful title, helpful description and a linked NuGet/Home issue
- [x] Added tests
- [x] Link to an issue or pull request to update docs if this PR changes settings, environment variables, new feature, etc. - https://github.com/NuGet/Home/issues/13963

kartheekp-ms committed:
Improve certificate fingerprint validation and tests

Enhance validation logic in `SignCommand` to check for SHA-1
fingerprints and throw an error for assembly versions 10+.
Update error messages to include error codes. Refactor
`ExecuteSignPackageTestWithCertificateFingerprintAsync` in
`DotnetSignTests` for better readability. Modify resource
strings to support new error message format.

kartheekp-ms committed:
Add version check to DotnetSign test for int assembly

Modified DotnetSign_SignPackageWithInsecureCertificateFingerprint_RaisesWarningAsync in DotnetSignTests.cs to include a conditional check based on the major version of the int assembly. If the major version is 10 or higher, the test asserts that the signing result is unsuccessful and that the errors contain the insecure certificate fingerprint code. If the major version is less than 10, the test retains the original behavior, asserting that the signing result is successful and that the output contains the insecure certificate fingerprint code.

jeffkl reviewed (dismissed):
You'll need to fix the compiler error

kartheekp-ms committed:
Remove unused namespaces from SignCommand.cs

The `using System.Reflection;` and `using System.Security.Cryptography.X509Certificates;` directives were removed from the `SignCommand.cs` file as they are no longer being used.

jgonz120 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx:
```diff
@@ -971,7 +971,8 @@ Non-HTTPS access will be removed in a future version. Consider migrating to 'HTT
   </data>
   <data name="SignCommandInvalidCertificateFingerprint" xml:space="preserve">
-    <value>Invalid value for '--certificate-fingerprint' option. The value must be a SHA-256, SHA-384, or SHA-512 certificate fingerprint (in hexadecimal).</value>
+    <value>{0}: Invalid value for '--certificate-fingerprint' option. The value must be a SHA-256, SHA-384, or SHA-512 certificate fingerprint (in hexadecimal).</value>
```
looks like you have a test failing regarding this message

kartheekp-ms replied:
Yeah, I have fixed the tests failing due to string mismatch.

kartheekp-ms committed:
Update exception message assertion in SignCommand test

The test method `SignCommandArgParsing_ThrowsAnExceptionForInvalidCertificateFingerpr` in `XplatSignTests.cs` has been updated to change the way it validates the exception message. Instead of asserting that the exception message is equal to a specific string, it now asserts that the exception message contains the `NuGetLogCode.NU3043` error code. This change likely makes the test more flexible and less dependent on the exact wording of the error message.

aortiz-msft commented:
@kartheekp-ms - This is a .NET SDK breaking change, so we need to follow the corresponding process.
